pkg/events/grpc/grpc.go

```diff
@@ -119,6 +135,59 @@ func (srv *EventsServer) Stream(req *ttnpb.StreamEventsRequest, stream ttnpb.Eve
 	}
 }
 
+var errStorageDisabled = errors.DefineFailedPrecondition("storage_disabled", "events storage is not not enabled")
+
+// FindRelated implements the EventsServer interface.
+func (srv *EventsServer) FindRelated(ctx context.Context, req *ttnpb.FindRelatedEventsRequest) (*ttnpb.FindRelatedEventsResponse, error) {
+	store, hasStore := srv.pubsub.(events.Store)
+	if !hasStore {
+		return nil, errStorageDisabled.New()
+	}
+	_, err := rights.AuthInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	evts, err := store.FindRelated(ctx, req.GetCorrelationID())
+	if err != nil {
+		return nil, err
+	}
+
+	var res ttnpb.FindRelatedEventsResponse
+
+	for _, evt := range evts {
+		evtProto, err := events.Proto(evt)
+		if err != nil {
+			log.FromContext(ctx).WithError(err).Warn("Failed to convert event to proto")
+			continue
+		}
+		isVisible, err := rightsutil.EventIsVisible(ctx, evt)
+		if err != nil {
+			log.FromContext(ctx).WithError(err).Warn("Failed to check event visibility")
+			continue
+		}
+		if isVisible {
+			res.Events = append(res.Events, evtProto)
+		} else {
+			res.Events = append(res.Events, &ttnpb.Event{
+				Name:        evtProto.Name,
+				Time:        evtProto.Time,
+				Identifiers: evtProto.Identifiers,
+				// Data is private
+				// CorrelationIDs is private
+				Origin: evtProto.Origin,
+				// Context is private
+				Visibility: evtProto.Visibility,
+				// Authentication is private
+				// RemoteIP is private
+				// UserAgent is private
+				UniqueID: evtProto.UniqueID,
+			})
```

Comment:
Point for discussion: Should (for example) a Gateway Owner be able to see end device events related to a downlink that they see on their gateway, even if they don't have rights on that application?

I think it would be useful to at least see _that_ something happened, and to see which application or end device caused some downlink, so that the gateway owner could come in contact with the application owner if something is wrong.

The current implementation therefore lets you see such events, without including unnecessary (private) fields.

Reply:
I think this needs a configuration toggle at least, since it allows people to 'fish' metadata to which they normally don't have access to. 

Our correlation IDs contain some entropy in them but we cannot guarantee our users have entropy into theirs.

Reply:
For private networks, it should really be no problem that end devices identifiers are available in gateway events. I think indeed that's very useful to see what's going on.

On the community network this might be more of an issue, but let's just try it and see how it goes.

Let's not add configuration, that needs documentation, default config, deployment templates etc, for things that we don't know is necessary to configure.